File: score-board-server/ScoreBoard.Domain/Enums/MatchState.cs
using System.Collections.Generic;

namespace ScoreBoard.Domain.Enums
{
    public class MatchState
    {
        private static readonly Dictionary<MatchStateEnum, MatchState> Dictionary = new Dictionary<MatchStateEnum, MatchState>();

        private MatchState(string name, MatchStateEnum matchStateEnum)
        {
            Name = name;
            MatchStateEnum = matchStateEnum;
            Dictionary.Add(matchStateEnum, this);
        }

        public string Name { get; }
        public MatchStateEnum MatchStateEnum { get; }

        public static IList<MatchState> GetAllMatchStates
        {
            get
            {
                IList<MatchState> allMatchState = new List<MatchState>();
                foreach (KeyValuePair<MatchStateEnum, MatchState> matchState in Dictionary)
                {
                    allMatchState.Add(matchState.Value);
                }
                return allMatchState;
            }
        }

        public static MatchState GetMatchState(MatchStateEnum matchStateEnum)
        {
            foreach (var (key, matchState) in Dictionary)
            {
                if (key == matchStateEnum)
                {
                    return matchState;
                }
            }
            return null;
        }

        public static MatchState GetMatchState(string matchStateString)
        {
            foreach (var (_, matchState) in Dictionary)
            {
                if (matchState.Name == matchStateString)
                {
                    return matchState;
                }
            }
            return null;
        }

        public static readonly MatchState Scheduled = new MatchState("Scheduled", MatchStateEnum.Scheduled);
        public static readonly MatchState Playing = new MatchState("Playing", MatchStateEnum.Playing);
        public static readonly MatchState Done = new MatchState("Done", MatchStateEnum.Done);
        public static readonly MatchState Interrupted = new MatchState("Interrupted", MatchStateEnum.Interrupted);
        public static readonly MatchState Aborted = new MatchState("Aborted", MatchStateEnum.Aborted);
    }

    public enum MatchStateEnum
    {
        Scheduled = 0,
        Playing = 1,
        Done = 2,
        Interrupted = 3,
        Aborted = 4
    }
}

File: score-board-client/src/app/models/match-state.ts
enum MatchStateEnum {
  Scheduled = 'Scheduled',
  Playing = 'Playing',
  Done = 'Done',
  Interrupted = 'Interrupted',
  Aborted = 'Aborted'
}

export class MatchState {
  private constructor(matchStateCode: string, matchStateEnum: MatchStateEnum, displayName: string) {
    this.matchStateCode = matchStateCode;
    this.matchStateEnum = matchStateEnum;
    this.displayName = displayName;
    MatchState.matchStates.push(this);
  }
  private static matchStates: Array<MatchState> = new Array<MatchState>();

  public static Scheduled = new MatchState('Scheduled', MatchStateEnum.Scheduled, 'Scheduled');
  public static Playing = new MatchState('Playing', MatchStateEnum.Playing, 'Playing');
  public static Done = new MatchState('Done', MatchStateEnum.Done, 'Done');
  public static Interrupted = new MatchState('Interrupted', MatchStateEnum.Interrupted, 'Interrupted');
  public static Aborted = new MatchState('Aborted', MatchStateEnum.Aborted, 'Aborted');

  matchStateCode: string;
  matchStateEnum: MatchStateEnum;
  displayName: string;

  static getMatchStates(): Array<MatchState> {
    return MatchState.matchStates;
  }

  static getMatchState(matchStateCode: string) {
    for (const matchState of MatchState.matchStates) {
      if (matchStateCode === matchState.matchStateCode) {
        return matchState;
      }
    }
    return null;
  }
}
